### PR TITLE
Tag box

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

File: Applications/Spire/Include/Spire/Ui/TagBox.hpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+      /**
+       * Signals that the tag at the index is being deleted.
+       * @param index The index of the deleted tag.
+       */
+      using DeleteSignal = Signal<void (int index)>;
+
+      /**
+       * Constructs a TagBox.
+       * @param list_model The list model which holds a list of tags.
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/TagBox.hpp:35]: Rename ```list_model``` to ```list``` and ```current_model``` to ```current```.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+  }
+}
+
+struct TagBox::PartialListModel : public ListModel {
+  std::shared_ptr<ListModel> m_source;
+
+  explicit PartialListModel(std::shared_ptr<ListModel> source)
+    : m_source(std::move(source)) {}
+
+  virtual int get_size() const {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:58]: Did you mean to make this an ```override```?

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+    return m_source->get_size() + 2;
+  }
+
+  virtual const std::any& at(int index) const {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:62]: See above.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+      return QValidator::State::Invalid;
+    }
+    if(index < m_source->get_size()) {
+      return m_source->set(index, value);
+    }
+    return QValidator::State::Acceptable;
+  }
+
+  connection connect_operation_signal(
+    const OperationSignal::slot_type& slot) const override {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:84]: Fix indentation.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+
+TagBox::TagBox(std::shared_ptr<ListModel> list_model,
+    std::shared_ptr<TextModel> current_model, QWidget* parent)
+    : QWidget(parent),
+      m_model(std::make_shared<PartialListModel>(std::move(list_model))),
+      m_focus_observer(*this),
+      m_overflow(TagBoxOverflow::WRAP),
+      m_tags_width(0),
+      m_list_item_gap(0) {
+  auto layout = new QHBoxLayout(this);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:97]: This is not used until line 122. Keep things local to avoid having to jump all over the place to understand how something works.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+  m_text_box = new TextBox(std::move(current_model));
+  set_style(*m_text_box, TEXT_BOX_STYLE(get_style(*m_text_box)));
+  m_text_box->installEventFilter(this);
+  m_list_view = new ListView(m_model,
+    std::bind_front(&TagBox::build_tag, this));
+  m_list_view->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  set_style(*m_list_view, LIST_VIEW_STYLE(get_style(*m_list_view)));
+  m_list_view_style_connection = connect_style_signal(*m_list_view,
+    std::bind_front(&TagBox::on_list_view_style, this));
+  for(auto i = 0; i < m_list_view->get_list()->get_size(); ++i) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:108]: Setting style like this is very fragile. Use a combinator instead, refer to CSS for how to style children:

https://developer.mozilla.org/en-US/docs/Web/CSS/Child_combinator

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+  auto input_box = make_input_box(m_list_view_container);
+  layout->addWidget(input_box);
+  proxy_style(*this, *input_box);
+  set_style(*this, INPUT_BOX_STYLE(get_style(*input_box)));
+  m_style_connection = connect_style_signal(*this, [=] { on_style(); });
+  m_ellipses_item = m_list_view->get_list_item(get_list()->get_size());
+  m_ellipses_item->hide();
+  setFocusProxy(m_text_box);
+  setFocusPolicy(Qt::StrongFocus);
+  m_line_edit = m_text_box->findChild<QLineEdit*>();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:130]: Remove the use of ```findChild<QLineEdit*>```, instead use ```TextBox::get_selection()```.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+const std::shared_ptr<TextModel>& TagBox::get_current() const {
+  return m_text_box->get_current();
+}
+
+bool TagBox::is_read_only() const {
+  return m_text_box->is_read_only();
+}
+
+void TagBox::set_read_only(bool is_read_only) {
+  m_text_box->set_read_only(is_read_only);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:150]: Look at other widgets to see how they handle this.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+    const DeleteSignal::slot_type& slot) const {
+  return m_delete_signal.connect(slot);
+}
+
+bool TagBox::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_text_box && event->type() == QEvent::KeyPress) {
+    auto& key_event = *static_cast<QKeyEvent*>(event);
+    switch(key_event.key()) {
+      case Qt::Key_Backspace:
+        if(m_line_edit->cursorPosition() == 0 && get_list()->get_size() > 0) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:171]: Use ```TextBox```'s selection model.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+    });
+    connect(tag, &QWidget::destroyed, [=] {
+      if(m_focus_observer.get_state() != FocusObserver::State::NONE) {
+        setFocus();
+      }
+    });
+    m_tags.emplace_back(tag);
+    return tag;
+  } else if(index == model->get_size() - 2) {
+    auto ellipses_box = make_label(QObject::tr("..."));
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:227]: Remove the ```QObject::```.

**Review by kranar — CHANGES_REQUESTED** [review]
Looks very good so far.

File: Applications/Spire/Source/Ui/TagBox.cpp (hunk 7 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/TagBox.cpp:130]: I have merged the latest commit, but I don't find TextBox::get_selection().

File: Applications/Spire/Source/Ui/TagBox.cpp (hunk 9 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/TagBox.cpp:171]: Since I didn't find the selection model, I used the highlight model to get the cursor position.

Comment (kranar) [reply Applications/Spire/Source/Ui/TagBox.cpp:171]: Yes yes, it's the highlight model not selection model. Good catch.

**Comment by shaobinspire** [comment]
I didn't find a way to get a QLineEdit from the TextBox, so I still used findChild<QLineEdit*> in the TagBox.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 11]
```diff
@@ -55,11 +62,11 @@ struct TagBox::PartialListModel : public ListModel {
   explicit PartialListModel(std::shared_ptr<ListModel> source)
     : m_source(std::move(source)) {}
 
-  virtual int get_size() const {
+  virtual int get_size() const override {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:65]: Remove the ```virtual```.

**Review by kranar — CHANGES_REQUESTED** [review]
Remove the use of ```QLineEdit*```.

File: Applications/Spire/Include/Spire/Ui/TagBox.hpp [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+      QMargins m_margins;
+      int m_tags_width;
+      int m_list_item_gap;
+      boost::signals2::scoped_connection m_highlight_connection;
+      boost::signals2::scoped_connection m_style_connection;
+      boost::signals2::scoped_connection m_list_view_style_connection;
+      boost::signals2::scoped_connection m_text_box_style_connection;
+      boost::signals2::scoped_connection m_focus_connection;
+
+      QWidget* build_tag(const std::shared_ptr<ListModel>& model, int index);
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/TagBox.hpp:89]: The convention in C++ is to name these functions ```make_xxx``` instead of ```build_xxx```. For example ```make_shared```, ```make_unique```, ```make_tuple```, etc...

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+  auto LIST_VIEW_STYLE(StyleSheet style) {
+    style.get(Any()).
+      set(ListItemGap(scale_width(4))).
+      set(ListOverflowGap(scale_width(3))).
+      set(Overflow::WRAP).
+      set(Qt::Horizontal);
+    auto& list_item_rule = style.get(Any() >> is_a<ListItem>());
+    auto list_item_style = LIST_ITEM_STYLE();
+    for(auto& rule : list_item_style.get_rules()) {
+      for(auto& property : rule.get_block()) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:42]: What is the goal of these nested for loops? Why are you applying a style to every rule instead of applying it to the ```Any()``` selector?

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+  }
+
+  QValidator::State set(int index, const std::any& value) override {
+    if(index < 0 || index >= get_size()) {
+      return QValidator::State::Invalid;
+    }
+    if(index < m_source->get_size()) {
+      return m_source->set(index, value);
+    }
+    return QValidator::State::Acceptable;
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:87]: Why does this return ```Acceptable```?

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+  }
+
+  connection connect_operation_signal(
+      const OperationSignal::slot_type& slot) const override {
+    return m_source->connect_operation_signal(slot);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:92]: This only works if ```PartialListModel``` owns ```m_source```. If it doesn't own ```m_source``` then you can not pass ```slot``` directly to ```m_source``` since when ```PartialListModel``` is destroyed, all those slots needs to be destroyed as well.

If ```PartialListModel``` owns ```m_source``` then there is no issue.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 16]
```diff
@@ -0,0 +1,513 @@
...
+  }
+  m_is_read_only = is_read_only;
+  m_text_box->set_read_only(m_is_read_only);
+  update_tags_read_only();
+  if(m_is_read_only) {
+    match(*this, ReadOnly());
+    match(*m_list_view, ReadOnly());
+  } else {
+    unmatch(*this, ReadOnly());
+    unmatch(*m_list_view, ReadOnly());
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:168]: Why is ```m_list_view``` ```ReadOnly```? This looks like an issue.

**Review by kranar — CHANGES_REQUESTED** [review]
This is very good work, I will need more time to complete my review but you can look at what I've reviewed so far.

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 17]
```diff
@@ -0,0 +1,514 @@
+#include "Spire/Ui/TagBox.hpp"
+#include <QHBoxLayout>
+#include <QKeyEvent>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ArrayListModel.hpp"
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:5]: Why was this included?

File: Applications/Spire/Source/Ui/TagBox.cpp [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+    remove_text_box_width_constraint();
+  }
+}
+
+void TagBox::reposition_list_view() {
+  if(m_overflow == TagBoxOverflow::ELIDE &&
+      m_focus_observer.get_state() == FocusObserver::State::FOCUS_IN) {
+    auto text_length = QFontMetrics(m_font).horizontalAdvance(
+      m_text_box->get_current()->get(), -1);
+    auto cursor_pos = [=] {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/TagBox.cpp:444]: Review the rules about how to properly capture in a lambda expression.

**Review by kranar — CHANGES_REQUESTED** [review]
Almost done, good work.
